File: Algorithm Design/Chromatic_Number_2_Python/graph.py
class Graph(): 
    def __init__(self, no_nodes): 
        self.N = no_nodes 
        self.graph = [[0 for i in range(no_nodes)] for j in range(no_nodes)] 
  
    #checks if it safe to color "node" with color "c"
    def safe_to_color(self, node, coloring_result, c): 
        for i in range(self.N): 
            if self.graph[node][i] == 1 and coloring_result[i] == c: 
                return False
        return True
    
    #utility function used to clean up the final coloring function
    def graph_color_utility_function(self, m, coloring_result, node): 
        if node == self.N: 
            return True
  
        for c in range(1, m+1): 
            if self.safe_to_color(node, coloring_result, c) == True: 
                coloring_result[node] = c 
                if self.graph_color_utility_function(m, coloring_result, node+1) == True: 
                    return True
                coloring_result[node] = 0
        return False

    #final coloring function, returns true if the coloring is complete
    def final_coloring_function(self, m): 
        coloring_result = [0] * self.N 
        if self.graph_color_utility_function(m, coloring_result, 0) == False:
            return False
        
        return True

File: Algorithm Design/Chromatic_Number_2_C/graph.c
#include "graph.h"

int safe_to_color(struct Graph g, int vertex, int coloring_result[], int color) {
    for (int i = 0; i < g.n; i++) {
        if (g.graph[vertex][i] == 1 && coloring_result[i] == color) {
            return 0;
        }
    }
    return 1;
}

int graph_color_utility_function(struct Graph g, int chromatic_number, int coloring_result[], int vertex) {
    if (vertex == g.n) {
        return 1;
    }
    for (int c = 1; c <= chromatic_number; c++) {
        if (safe_to_color(g, vertex, coloring_result, c)) {
            coloring_result[vertex] = c;
            if (graph_color_utility_function(g, chromatic_number, coloring_result, vertex + 1)) {
                return 1;
            }
            coloring_result[vertex] = 0;
        }
    }
    return 0;
}

int final_coloring_function(struct Graph g, int chromatic_number, int coloring_result[]) {
    for (int i = 0; i < g.n; i++) {
        coloring_result[i] = 0;
    }
    if (!graph_color_utility_function(g, chromatic_number, coloring_result, 0))
        return 0;
    return 1;
}

